feat: 출결 상태 수정 API 구현 및 출석 점수 자동 업데이트 구현

bugoverdose opened the pull request:
closes #60

Attendance 도메인 추가된 컬럼들
- 비고(note): 해당 출결 상태에 대한 전반적인 설명
- 기타 점수(extraScoreNote) : "운영지원"과 같은 가산점/감점에 대한 설명
- 점수 변화(scoreChanged) : 지각, 가산점 등을 전부 고려했을 때, 해당 출결로 인한 점수 변화. 해당 점수 변화들을 합산하고 100에서 뺀게 해당 기수 회원의 최종 점수

## 구현한 기능

1. `PUT /api/attendance/{id}/status` API 구현

  - 특정 회원의 회원의 출결 상태와 함께 비고(ex. 지각에 대한 원인 등)를 수정해주는 기능

2. `POST /api/attendance/{id}/additional-score` API 구현

  - 운영지원 등의 이유로 기타 점수를 부여하는 기능

3. **출석 상태 및 출결 점수 수정시, 해당 기수회원의 총 출결 점수도 자동으로 동기화되도록 구현**

- 일반 회원이 QR 코드를 찍어서 출석/지각을 하게 되면 점수 동기화
- 출석 종료에 의해 무단 결석 처리되었을 때에도 점수 동기화
- 관리자가 출석 상태 수정시에도 점수 동기화
- 관리자가 가산점/감점 부여했을 때에도 점수 동기화

## 유의할 점

- 특정 출결에 대한 점수 변화(`scoreChanged`) 컬럼 값은 덮어씌워지지 않고 무조건 **일정 값만큼 더해지거나 감소**함. 
  - 출결 상태를 변경할 경우, 기존 출결 상태의 패널티 점수와 변경하려는 출결 상태의 패널티 점수를 비교한 변화값만큼 값을 더하거나 뺌. (not 덮어씀)
  - 굳이 이렇게 하는 이유는 [운영지원 +5]과 같은 가산점 부여 이후에, 출결 상태를 수정했을 때 가산점이 지워지지 않도록 하기 위함.
  - 헷갈릴 수 있을 것 같은데 일단은 차악인것 같아서 이렇게 구현함

## 공유사항

- 인수테스트 없이 일단 서비스 & 도메인 테스트만 구현 완료

bugoverdose committed:
feat: Attendance 엔티티 수정

bugoverdose committed:
feat: 개별 출석 도메인에 대한 점수 변화 메서드 구현

bugoverdose committed:
feat: 관리자에 의한 출석 상태 수정 API 구현

bugoverdose committed:
feat: 관리자에 의한 출석 가산점/감점 부여 API 구현

bugoverdose committed:
test: Pending 상태에서 출석 도메인 변화가 시작하도록 설정

bugoverdose committed:
Merge branch 'develop' into feat/additional-attendance-info

bugoverdose committed:
Merge branch 'develop' into feat/additional-attendance-info

xxeol2 commented on src/main/kotlin/nexters/admin/domain/attendance/Attendance.kt:
```diff
@@ -9,15 +9,24 @@ class Attendance(
+        @Column(name = "attendance_status", nullable = false)
+        var attendanceStatus: AttendanceStatus,
+
+        @Column(name = "extra_score_note")
+        var extraScoreNote: String? = "",
```
nullable인데 null대신 빈문자열로 초기화해준 이유가 있을까요?!!
어제 얘기한 것 같기두 하구.. 기억이 잘 안나서 물어봅니다 ㅎㅎㅎ

bugoverdose replied:
빈문자열로 해도 되고 null로 해도 되니 아무렇게나 하자?라고 얘기됐던 것 같은데, 그냥 가급적 null은 안보내는게 좋을 것 같아서 빈문자열을 넣은 느낌?

kth990303 replied:
맞습니다~

bugoverdose committed:
Merge branch 'develop' into feat/additional-attendance-info

bugoverdose committed:
style: 불필요한 개행 제거

xxeol2 reviewed (approved):
LGTM!!

bugoverdose committed:
feat: 기수회원 도메인의 점수 동기화 메서드 구현

- 가산점 및 감점 목록을 기준으로 기수회원의 출결 상태 계산 및 수정

bugoverdose committed:
feat: 출석 상태 및 출결 점수 수정시, 기수회원의 점수 자동 동기화 기능 구현

- 일반 회원의 QR 코드 기반 출석시, 점수 동기화
- 출석 종료에 의해 무단 결석 처리되었을 때에도 점수 동기화
- 관리자가 출석 상태 수정시에도 점수 동기화
- 관리자가 가산점 부여했을 때에도 점수 동기화

bugoverdose committed:
fix: develop 브랜치 머지 후 충돌 해결

kth990303 commented on src/main/kotlin/nexters/admin/controller/attendance/AttendanceDtos.kt:
```diff
@@ -22,6 +22,16 @@ data class CurrentQrCodeResponse(
 }
 
+data class ExtraAttendanceScoreChangeRequest(
+        val extraScoreChange: Int,
```
빈 문자열이 들어올 거 같긴 한데 nullable해도 괜찮을 거 같아서 한번 코멘트 남겨봅니다
<img width="751" alt="image" src="https://user-images.githubusercontent.com/57135043/221584665-1c1293bb-eb70-48db-bdcf-56d6e10e689e.png">

kth990303 commented on src/test/kotlin/nexters/admin/domain/attendance/AttendanceTest.kt:
```diff
@@ -0,0 +1,33 @@
+    fun `QR 코드 혹은 관리자에 의한 출석 상태 수정시 기존 출석 상태를 기준으로 점수 변화`() {
+        val attendance = createNewPendingAttendance()
+
+        attendance.updateStatusByAdmin(AttendanceStatus.AUTHORIZED_ABSENCE, "통보 결석")
+        attendance.updateStatusByQr(AttendanceStatus.TARDY)
```
관리자가 통보결석으로 처리했다 하더라도, 회원이 지각으로 처리하면 해당 상태로(무조건 최신꺼로) 변경되는거군요. 이해한게 맞을까요?
좋은 거 같아요 👍

bugoverdose replied:
예스. 상태변화 작업이 여러번 있으면 무조건 최종껄로 반영되는 느낌.

kth990303 reviewed (approved):
이제 develop 브랜치와의 충돌을 해결하고 인수테스트를 짜봅시다~ 하하하

bugoverdose committed:
feat: 관리자의 출석 점수 변경에 대한 인수테스트 추가

bugoverdose committed:
refactor: 개행 수정

bugoverdose commented on src/test/kotlin/nexters/admin/acceptance/AttendanceAcceptanceTest.kt:
```diff
@@ -67,22 +52,12 @@ class AttendanceAcceptanceTest : AcceptanceTest() {
-        } Extract {
-            `as`(AttendanceSessionResponses::class.java)
-        }
+        val attendanceSessionResponses = 세션_출석_조회(adminToken, sessionId)
```
아래쪽 인수테스트들의 given/when 부분에서 사용하기 위해 AcceptanceFixtures로 이동

kth990303 replied:
픽스쳐 굳굳
근데 해당 테스트에서 endpoint나 statusCode 검증을 위해 픽스쳐로 변환하진 않아도 될 거 같은 느낌?

kth990303 reviewed (approved):
인수테스트 추가 👍
